cilium/CFP-23380-ordered-policy.md

```diff
@@ -0,0 +1,329 @@
+# CFP-23380: Ordered network policy
+
+**SIG: sig-policy**
+
+**Begin Design Discussion:** 2024-07-01
+
+**Cilium Release:** 1.17
+
+**Authors:** Casey Callendrello <cdc@isovalent.com>, You, and you too.
+
+## Summary
+
+Add explicit, user-supplied rule ordering to the policy engine.
+
+## Motivation
+
+Most firewall engines present an _iterative_ interface to their users. Users provide an _ordered_ list of rules. In comparison, Cilium policy is _commutative_, which is to say, the ordering of rules does not matter.
+
+Iterative policy languages are more expressive. For example, consider the rules:
+
+> 1. `env=qa, role=grafana`: allow access to `env=prod, role=artifacts`
+> 2. `env=qa`: deny access to `env=prod`
+
```

Comment:
I don't know if there's a good place to discuss this, but IMO cloud-native has substituted iteration for label selectors to express selection. Combining both is significantly complicating the solution over traditional iterative approaches, because you _can't_ just look at the priorities, you still have to know how the selectors apply to workloads based on their labels. This is part of what makes implementing this proposal so tricky, and what will make debugging environments with this feature also very tricky.

Reply:
Come to think of it, we should probably highlight that debuggability impact in its own impact section.